### ISSUE TITLE
incorrect handling of promise resolve vs method return

### ISSUE THREAD

**Issue by vladmandic** [microsoft/TypeScript#51624]
TypeScript does not treat code after call to `resolve()` inside promise as unreachable thus leading to unnecessary errors being reported  
This is unlike any code after that occurs after call to `return` where TypeScript behavior is as expected

```ts
async function test() {
  return new Promise((resolve) => {
    let testVariable: number | undefined;
    if (!testVariable) resolve(true); // this resolves immediately as variable is not defined
    // typescript continues with checks and next line results in error although clearly there is no error due to early exit with `resolve()`
    // ts(18048) error 'testVariable' is possibly 'undefined'
    if (testVariable > 0) console.log(testVariable);
    // now lets see how it handles return calls
    if (!testVariable) return;
    // no error at all altough return call here was completely unnecessary
    if (testVariable > 0) console.log(testVariable);
  });
}
```

Tested with TypeScript 4.9, although older versions are also affected

**Comment by vladmandic** [comment]
not sure why github states i've added #duplicate label as i have done no such thing?!

**Comment by MartinJohns** [comment]
Because you used the wrong issue template. You used the issue template for wrong types after/ in callbacks, which is a very common duplicate.

**Comment by fatcerberus** [comment]
@vladmandic for the record this is not a bug - ‘resolve()` is not an early return and so the subsequent code is indeed reachable.
